activerecord/lib/active_record/connection_adapters/metadata_registry.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require "digest/md5"
+
+module ActiveRecord
+  module ConnectionAdapters # :nodoc:
+    module MetadataRegistry
+      REGISTRY = {}
+
+      class << self
+        def fetch(type, *args)
+          key = Digest::MD5.hexdigest(Marshal.dump([type.name, args]))
+          REGISTRY[key] ||= yield.freeze
+        end
+      end
+
+      module Concern
+        def new(*args)
+          MetadataRegistry.fetch(self, *args) { super }
```

Comment:
Does Ruby optimize splatting and unsplatting these days? How many arrays does this create?

Since `fetch` isn't meant to be public we could make the signature `def fetch(type, args)`.